src/napari/layers/_layer_actions.py

```diff
@@ -56,7 +56,11 @@ def _convert(ll: LayerList, type_: str) -> None:
     for lay in list(ll.selection):
         idx = ll.index(lay)
         if isinstance(lay, Shapes) and type_ == 'labels':
-            data = lay.to_labels()
+            # lay.scale is added to the label world extent to account for the virtual width
+            # of each data index (nominally 1.0 for unscaled images)
+            data = lay.to_labels(
+                labels_shape=lay.world_to_data(ll.extent.world[-1] + lay.scale)
+            )
```

Comment:
If I understand this comment correctly, what you want here is the `ll._extent_world_augmented` (extent which accounts for data voxel/point sizes).

```suggestion
            ll_shape = ll._extent_world_augmented[1] - ll._extent_world_augmented[0]
            data = lay.to_labels(
                labels_shape=lay.world_to_data(ll_shape)
            )
```

Looks like there's no need for converting to int because this already happens in the `to_labels()` method.

Reply:
no matter what, I sympathize @rwkozar, this part of the code base is really hard for me to understand, and I've mucked about quite a few times. 🙃

Reply:
Thanks @brisvag ! I knew there had to be a better way. @rwkozar was a trooper for working through all the caveats in synthetic tests and a real viewer. I will try to find some time to look more carefully -- i'm frantically trying to catch up on a week worth of missed work right now.

Reply:
I played with this some more and I think the suggestion here is close, but then I ran into issues, but they're on main:
https://github.com/napari/napari/issues/8115
I will try to clean up what I have and make suggestion.

Reply:
@psobolewskiPhD I think it would be best if you resolve Lorenzo's comment.  I could make the change but could not really tell if it is the right/best solution.

Reply:
Will we have a fix on this soon, or should we bump to next release?

Reply:
@TimMonko Without tracking down what world_to_data and _extent_world_augmented do exactly, it would be hard for me to assess which is the best approach.  We did test what's in the code now quite a bit and it worked for all the cases as long as the image stays in positive coordinates.  If this is closed, it would be good to add an issue for addressing negative coordinates, as noted in the PR description.

Reply:
I'll try to sit down with this soon, because it would take me a bit to sort through. I totally agree with you, the naming of some of those objects are really difficult, and I have to deeply think about that section of code every time. 

As for me, if this isn't a regression in any way (i.e., it may not fully work in every case, but is strictly an improvement in the positive coords case) then I thinks its useful to merge as laid out, and we work on a fix in a follow-up PR. 
I know @psobolewskiPhD has been quite busy recently, but he may be able to circle back and support this over the coming weekend.

Reply:
Based on my involvement, that would be my recommended path too.

Reply:
I was going to see what that update would do, but when I tried to `Convert to Labels`, I got a program error (`ValueError: axis_labels=('axis -2', 'axis -1') must have length ndim=3.`).  I also tried doing the conversion on the main branch and running the program with vu, but got the same error both times.  It could just be something I have configured wrong on my system, but may be worth checking out.

Reply:
Ok, again apologies for the delay, i've just been swamped.
I tested this a few times. So with @brisvag this works when there is no translation involved, but if translation is involved then it fails, because the translation offset is not taken into account.
Simplest test: open astronaut and use the transform tool to scale from the upper left towards lower right (scale + translation), then add Shapes, draw and Shape, and convert to labels. The labels is all black.
The current code here handles that case, with the caveat that the labels layer extends to (0, 0).

But using `ll._extent_world_augmented[-1]` does clean up the `lay.scale` part.

Reply:
I think https://github.com/napari/napari/issues/8115 is also showing us that this whole method needs some rethinking.
Also, it appears that new Shapes layers don't inherit `Layer.affine` or `Layer.translate`, just `Layer.scale` which makes it hard to handle this using those properties!

Reply:
Ok, i'm back at this. I think the root cause of this issue is much more complex. With this suggestion things work well for the simple cases that fail on main, so this is a strict improvement. Let's go with that.
scale works, but not translation.
The real issue is that the to_labels code was designed for making thumbnails and doesn't handle a lot of cases well, see also https://github.com/napari/napari/issues/8115#issuecomment-3172296726